ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -238,6 +292,190 @@ VALUE cpu_and_wall_time_collector_sample(VALUE self_instance) {
+}
+
+// This function gets called when Ruby has finished running the Garbage Collector on the current thread.
+// It updates the per_thread_context of the current thread to include the current cpu/wall times, to be used to later
+// create a stack sample that blames the cpu/wall time spent from the start of garbage collector work until now.
+//
+// Safety: This function gets called while Ruby is doing garbage collection. While Ruby is doing garbage collection,
+// *NO ALLOCATION* is allowed. This function, and any it calls must never trigger memory or object allocation.
+// This includes exceptions and use of ruby_xcalloc (because xcalloc can trigger GC)!
+//
+// Assumption 1: This function is called in a thread that is holding the Global VM Lock. Caller is responsible for enforcing this.
+void cpu_and_wall_time_collector_on_gc_finish(VALUE self_instance) {
+  struct cpu_and_wall_time_collector_state *state;
+  if (!rb_typeddata_is_kind_of(self_instance, &cpu_and_wall_time_collector_typed_data)) return;
+  // This should never fail the the above check passes
+  TypedData_Get_Struct(self_instance, struct cpu_and_wall_time_collector_state, &cpu_and_wall_time_collector_typed_data, state);
+
+  struct per_thread_context *thread_context = get_context_for(rb_thread_current(), state);
+
+  // If there was no previously-existing context for this thread, we won't allocate one (see safety). We keep a metric for
+  // how often this happens -- see on_gc_start.
+  if (thread_context == NULL) return;
+
+  if (thread_context->gc_tracking.cpu_time_at_start_ns == INVALID_TIME &&
+    thread_context->gc_tracking.wall_time_at_start_ns == INVALID_TIME) {
+    // If this happened, it means that on_gc_start was either never called for the thread OR it was called but no thread
+    // context existed at the time. The former can be the result of a bug, but since we can't distinguish them, we just
+    // do nothing.
+    return;
+  }
+
+  // Here we record the wall-time second and in on_gc_start we record it first to avoid having wall-time be slightly < cpu-time
+  thread_context->gc_tracking.cpu_time_at_finish_ns = cpu_time_now_ns(thread_context);
+  thread_context->gc_tracking.wall_time_at_finish_ns = wall_time_now_ns(DO_NOT_RAISE_ON_FAILURE);
+}
+
+// This function gets called shortly after Ruby has finished running the Garbage Collector.
+// It creates a new sample including the cpu and wall-time spent by the garbage collector work, and resets any
+// GC-related tracking.
+//
+// Specifically, it will search for thread(s) which have gone through a cycle of on_gc_start/on_gc_finish
+// and thus have cpu_time_at_start_ns, cpu_time_at_finish_ns, wall_time_at_start_ns, wall_time_at_finish_ns
+// set on their context.
+//
+// Assumption 1: This function is called in a thread that is holding the Global VM Lock. Caller is responsible for enforcing this.
+// Assumption 2: This function is allowed to raise exceptions. Caller is responsible for handling them, if needed.
```

Comment:
I think it's worth explicitly mentioning that this method is allowed to allocate memory, in contrasts with `cpu_and_wall_time_collector_on_gc_finish` and `cpu_and_wall_time_collector_on_gc_start`.

Reply:
Agreed! Added in d991cf72f34f7a34db23f6a387256ded1d4966a5